### PR TITLE
Add non-root user to run temporal server

### PR DESCRIPTION
**Pull request by sagardeo2105**
Adding non-root user for temporal server

Currently the temporal server only runs as root user, which is not the recommended way of running containers on kubernetes. 

I tested this change by deploying on test kubernetes environment

**Potential risks**
There will be a backward compatibility issue.

**Is hotfix candidate?** No

### CODE REVIEW

**Comment by SiweiWang** [comment]
addresses https://github.com/temporalio/temporal/issues/1263

File: Dockerfile [hunk 1]
```diff
@@ -18,19 +18,22 @@ RUN make bins
...
 COPY docker/entrypoint.sh /etc/temporal/entrypoint.sh
 COPY docker/start-temporal.sh /etc/temporal/start-temporal.sh
-
 COPY --from=temporal-builder /temporal/tctl /usr/local/bin
 COPY --from=temporal-builder /temporal/temporal-server /usr/local/bin
 
+RUN chown -R temporal:temporal /etc/temporal
+USER temporal
+
+ENTRYPOINT ["./entrypoint.sh"]
```
Comment (alexshtin) [inline Dockerfile:36]: Why did you move `ENTRYPOINT`?

File: docker/base-images/base-server.Dockerfile [hunk 2]
```diff
@@ -22,6 +22,9 @@ RUN apk add --update --no-cache \
     curl \
     vim
 
+RUN addgroup -g 1000 temporal
+RUN adduser -u 1000 -G temporal -h /home/temporal -D temporal
+
```
Comment (alexshtin) [inline docker/base-images/base-server.Dockerfile:27]: Why did you put these lines to base image? The idea of base images was that there is no Temporal specific stuff but only preinstalled 3rd party libs.

File: Dockerfile [hunk 3]
```diff
@@ -18,19 +18,22 @@ RUN make bins
...
 
 COPY config/dynamicconfig /etc/temporal/config/dynamicconfig
 COPY docker/config_template.yaml /etc/temporal/config/config_template.yaml
 COPY docker/entrypoint.sh /etc/temporal/entrypoint.sh
 COPY docker/start-temporal.sh /etc/temporal/start-temporal.sh
-
 COPY --from=temporal-builder /temporal/tctl /usr/local/bin
 COPY --from=temporal-builder /temporal/temporal-server /usr/local/bin
 
+RUN chown -R temporal:temporal /etc/temporal
```
Comment (alexshtin) [inline Dockerfile:33]: Shouldn't we change `WORKDIR` to `/home/temporal` instead?

File: docker/base-images/base-server.Dockerfile (hunk 2 above)
Comment (sagardeo2105) [reply docker/base-images/base-server.Dockerfile:27]: I will move them to the DockerFile

File: Dockerfile (hunk 1 above)
Comment (sagardeo2105) [reply Dockerfile:36]: Why are we copying files after the `ENTRYPOINT`?

File: docker/base-images/base-server.Dockerfile (hunk 2 above)
Comment (sagardeo2105) [reply docker/base-images/base-server.Dockerfile]: This is moved to main dockerfile

File: Dockerfile (hunk 1 above)
Comment (sagardeo2105) [reply Dockerfile:41]: @alexshtin

Comment (alexshtin) [reply Dockerfile:41]: yes, why did you move it from line 24 to 41?

**Review by underrun — APPROVED** [review]

File: Dockerfile (hunk 3 above)
Comment (alexshtin) [reply Dockerfile:38]: I think this is a right way to do it. Remove this line and change `/etc/temporal` to `/home/temporal` everywhere in `Dockerfile`.

File: Dockerfile (hunk 1 above)
Comment (alexshtin) [reply Dockerfile:41]: Please move it back, and lines 39 to 23.

Comment (sagardeo2105) [reply Dockerfile:42]: @alexshtin why would we want to copy files after ENTRYPOINT? 
It does not make sense. 

Can you please elaborate on why it is the way it is?

**Comment by sagardeo2105** [comment]
@alexshtin Sent you the invite.

**Review by alexshtin — APPROVED** [review]

**Comment by alexshtin** [comment]
I finally merged it. We decided not to include it in the release I did today (1.12), but will be in the next one. @sagardeo2105, sorry it took 21 day to merge.

**Comment by sagardeo2105** [comment]
https://github.com/temporalio/helm-charts/pull/221
We need to merge this as well :) 
Else the helm charts won't work.

Can you help review and merge that as well @alexshtin

**Comment by hazcod** [comment]
Also, please set the default user to a numeric ID and not a username. Else Kubernetes will throw an error;
```
Error: container has runAsNonRoot and image has non-numeric user (temporal), cannot verify user is non-root (pod: "temporal-c747c9478-7h4gt_default(5cfd4679-1bd0-43b0-8ae9-032a04ea8789)", container: temporal)
```

**Comment by alexshtin** [comment]
temporalio/helm-charts#221 is merged! And I double checked that is uses numeric Id, not `temporal`.

File: Dockerfile [hunk 4]
```diff
@@ -17,16 +17,20 @@ RUN make bins
...
-
 WORKDIR /etc/temporal
-
 ENV TEMPORAL_HOME /etc/temporal
 ENV SERVICES "history:matching:frontend:worker"
 EXPOSE 6933 6934 6935 6939 7233 7234 7235 7239
 
+# TODO (alex): switch WORKDIR to /home/temporal and remove "mkdir" and "chown" calls.
+RUN addgroup -g 1000 temporal
+RUN adduser -u 1000 -G temporal -D temporal
```
Comment (jrsmiley) [inline Dockerfile:27]: User ID 1000 may conflict with the uid of a real user.  Some Docker implementations require a uid:gid to be passed to the container and those are frequently the uid:gid if the actual user running the container.  Suggest using "nobody" as the uid to run temporal under.

Comment (hazcod) [reply Dockerfile:27]: @jrsmiley: not sure what you mean here, since the base image never adds a user and `1000` is popular as the first UID to be used.

Comment (jrsmiley) [reply Dockerfile:27]: In some Docker implementations designed to avoid allowing root, the user must input a uid:gid that the container will run with.  That means that all uid:gids are possible except for "nobody" and "nogroup".

Comment (hazcod) [reply Dockerfile:27]: I believe the better practice there is user namespace remapping.
Changing internal UIDs/GIDs is a workaround IMO.
